File: Easy/1160_Find_Words_That_Can_Be_Formed_by_Characters/Solution.java


public class Solution {
    public int countCharacters(String[] words, String chars) {
        int[] freq = new int[26];

        for (char ch : chars.toCharArray()) {
            freq[ch - 'a']++;
        }

        int ans = 0;

        for (String word : words) {
            boolean good = true;
            int[] temp = freq.clone();

            for (char ch : word.toCharArray()) {
            
                if (--temp[ch - 'a'] < 0) {
                    good = false;
                    break;
                }
            }

            if (good) ans += word.length();
        }

        return ans;
    }

    public static void main(String[] args) {
        Solution solution = new Solution();

        // test cases 1
        String[] words1 = {"cat", "bt", "hat", "tree"};
        String chars1 = "atach";
        System.out.println(solution.countCharacters(words1, chars1)); // Output: 6

        // test cases 2
        String[] words2 = {"hello", "world", "leetcode"};
        String chars2 = "welldonehoneyr";
        System.out.println(solution.countCharacters(words2, chars2)); // Output: 10
        
    }
}

File: Easy/1160_Find_Words_That_Can_Be_Formed_by_Characters/Solution.py
from typing import List, Optional


class Solution:
    def countCharacters(self, words: List[str], chars: str) -> int:
        freq = [0] * 26

        for i in chars:
            freq[ord(i) - ord('a')] += 1

        ans = 0

        for word in words:
            temp = freq.copy()
            good = True

            for ch in word:
                if temp[ord(ch) - ord('a')] > 0:
                    temp[ord(ch) - ord('a')] -= 1
                else:
                    good = False
                    break
                
            if (good):
                ans += len(word)

        return ans


if __name__ == "__main__":
    solution = Solution()

    # test cases 1
    words1 = ["cat", "bt", "hat", "tree"]
    chars1 = "atach"
    print(solution.countCharacters(words1, chars1))  # Output: 6
    
    # test cases 2
    words2 = ["hello", "world", "leetcode"]
    chars2 = "welldonehoneyr"
    print(solution.countCharacters(words2, chars2))  # Output: 10
    
    